packages.nix

```diff
@@ -1,32 +1,47 @@
 let
-# Pinning to revision f54322490f509985fa8be4ac9304f368bd8ab924 
-# - cln v24.02.1
-# - lnd v0.17.4-beta
-# - bitcoin v26.0
-# - elements v23.2.1
+    fetchNixpkgs = rev: fetchTarball "https://github.com/NixOS/nixpkgs/archive/${rev}.tar.gz";
+    # Pinning to revision f54322490f509985fa8be4ac9304f368bd8ab924 
+    # - cln v24.02.1
+    # - lnd v0.17.4-beta
+    # - bitcoin v26.0
+    # - elements v23.2.1
+    rev1 = "f54322490f509985fa8be4ac9304f368bd8ab924";
+    nixpkgs1 = fetchNixpkgs rev1;
+    pkgs1 = import nixpkgs1 {};
 
-rev = "f54322490f509985fa8be4ac9304f368bd8ab924";
-nixpkgs = fetchTarball "https://github.com/NixOS/nixpkgs/archive/${rev}.tar.gz";
-pkgs = import nixpkgs {};
+    # Override priority for bitcoin as /bin/bitcoin_test will
+    # confilict with /bin/bitcoin_test from elementsd.
+    bitcoind = (pkgs1.bitcoind.overrideAttrs (attrs: {
+        meta = attrs.meta or {} // {
+            priority = 0;
+        };
+    }));
+    # lwk: init at 0.3.0 #292522
+    # https://github.com/NixOS/nixpkgs/pull/292522/commits/2b3750792b2e4b52f472b6e6d88a6b02b6536c43
+    rev2 = "2b3750792b2e4b52f472b6e6d88a6b02b6536c43";
+    nixpkgs2 = fetchNixpkgs rev2;
+    pkgs2 = import nixpkgs2 {};
+    # blockstream-electrs: init at 0.4.1 #299761
+    # https://github.com/NixOS/nixpkgs/pull/299761/commits/680d27ad847801af781e0a99e4b87ed73965c69a
+    rev3 = "680d27ad847801af781e0a99e4b87ed73965c69a";
+    nixpkgs3 = fetchNixpkgs rev3;
+    pkgs3 = import nixpkgs3 {};
+    blockstream-electrs = pkgs3.blockstream-electrs.overrideAttrs (oldAttrs: {
+        cargoBuildFlags = [ "--features liquid" "--bin electrs" ];
+    });
```

Comment:
There are also nix flakes for lwk and blockstream-electrs, I am curious to know if it was unknown or if it's preferred to avoid flakes

Reply:
Thank you for confirming this.
I know that flake exists and I'm not trying to avoid installing it.

Currently in peerswap we use `nix-shell` to do integration tests.
I think with flake it is needed to be able to prepare an environment with `nix develop`.
The reason why flake has not been introduced is that there are [no packages in blockstream-electrs](https://github.com/Blockstream/electrs/blob/new-index/flake.nix#L62) built with `--features liquid`.

Is there a better way?
Would it be better to add `binLiquid` to the electrs package?

Reply:
Thank you for confirming this.  
I know that flake exists and I'm not trying to avoid installing it.

I don't know enough about nix but would it be better to add `binLiquid` to the package and use it for peerswap devShell?
I created [draft PR](https://github.com/Blockstream/electrs/pull/83) of this.

I also seem to get build errors.
I think it should be fixed.

```sh
nix run .#blockstream-electrs-liquid -- --network liquid
error: builder for '/nix/store/nkkv9v3r310p5gp09zv3hmn29y63mia3-electrs-deps-0.4.1.drv' failed with exit code 101;
       last 25 log lines:
       >     Checking futures-channel v0.3.30
       >     Checking which v4.4.2
       >    Compiling bitcoind v0.34.1
       >     Checking rand v0.4.6
       >     Checking rustc-demangle v0.1.23
       >     Checking iana-time-zone v0.1.60
       >     Checking unicode-width v0.1.11
       >     Checking tower-service v0.3.2
       >     Checking winapi v0.2.8
       >     Checking option-ext v0.2.0
       >    Compiling bitcoin_hashes v0.10.0
       >     Checking base64-compat v1.0.0
       >     Checking dirs-sys v0.4.1
       >     Checking hyper v0.14.28
       >     Checking bitcoincore-rpc-json v0.18.0
       >    Compiling elementsd v0.9.1
       > error[E0425]: cannot find function `download_filename` in this scope
       >   --> /nix/store/n1a9vx4rppg99v909c7hw2jc5d1yj7f2-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/elementsd-0.9.1/build.rs:59:33
       >    |
       > 59 |         let download_filename = download_filename();
       >    |                                 ^^^^^^^^^^^^^^^^^ not found in this scope
       >
       > For more information about this error, try `rustc --explain E0425`.
       > error: could not compile `elementsd` (build script) due to previous error
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix-store -l /nix/store/nkkv9v3r310p5gp09zv3hmn29y63mia3-electrs-deps-0.4.1.drv'.
error: 1 dependencies of derivation '/nix/store/qpdjg9rmj3b9w33l50gmg4jdl8671738-electrs-0.4.1.drv' failed to build
```

Reply:
Which commit are you using?

current new-index branch seems fine to me

```bash
$ git log | head -n 1
commit efc1fec8b0f96b5663d7257a0c2cffd8ef143219
$ nix run .#blockstream-electrs-liquid -- --network liquid
Config { log: StdErrLog { verbosity: Error, quiet: false, show_level: true, timestamp: Off, modules: [], writer: "stderr", color_choice: Auto, show_module_names: false }, network_type: Liquid, db_path: "./db/liquid", daemon_dir: "/home/casatta/.bitcoin/liquidv1", blocks_dir: "/home/casatta/.bitcoin/liquidv1/blocks", daemon_rpc_addr: 127.0.0.1:7041, cookie: None, electrum_rpc_addr: 127.0.0.1:51000, http_addr: 127.0.0.1:3000, http_socket_file: None, monitoring_addr: 127.0.0.1:34224, jsonrpc_import: false, light_mode: false, address_search: false, index_unspendables: false, cors: None, precache_scripts: None, utxos_limit: 500, electrum_txs_limit: 500, electrum_banner: "Welcome to electrs-esplora 0.4.1", electrum_rpc_logging: None, parent_network: Bitcoin, asset_db_path: None }

```

Reply:
I'm using the same commit.
There seems to be an error in my environment.

```sh
$ git log | head -n 1
commit efc1fec8b0f96b5663d7257a0c2cffd8ef143219

$ nix run .#blockstream-electrs-liquid -- --network liquid
error: builder for '/nix/store/nkkv9v3r310p5gp09zv3hmn29y63mia3-electrs-deps-0.4.1.drv' failed with exit code 101;
       last 25 log lines:
       >     Checking rand v0.4.6
       >     Checking tower-service v0.3.2
       >     Checking option-ext v0.2.0
       >     Checking winapi v0.2.8
       >     Checking rustc-demangle v0.1.23
       >     Checking unicode-width v0.1.11
       >    Compiling bitcoin_hashes v0.10.0
       >     Checking iana-time-zone v0.1.60
       >     Checking bitcoincore-rpc-json v0.18.0
       >     Checking base64-compat v1.0.0
       >     Checking chrono v0.4.34
       >     Checking textwrap v0.11.0
       >     Checking rand v0.3.23
       >     Checking hyper v0.14.28
       >     Checking dirs-sys v0.4.1
       >    Compiling elementsd v0.9.1
       > error[E0425]: cannot find function `download_filename` in this scope
       >   --> /nix/store/n1a9vx4rppg99v909c7hw2jc5d1yj7f2-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/elementsd-0.9.1/build.rs:59:33
       >    |
       > 59 |         let download_filename = download_filename();
       >    |                                 ^^^^^^^^^^^^^^^^^ not found in this scope
       >
       > For more information about this error, try `rustc --explain E0425`.
       > error: could not compile `elementsd` (build script) due to previous error
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/nkkv9v3r310p5gp09zv3hmn29y63mia3-electrs-deps-0.4.1.drv'.
error: 1 dependencies of derivation '/nix/store/qpdjg9rmj3b9w33l50gmg4jdl8671738-electrs-0.4.1.drv' failed to build
```

Reply:
can you try a `nix flake update` before `nix run` I am thinking to update the `flake.lock` anyway...

Reply:
Hmmm, doesn't seem to solve the problem.

```sh
$ nix flake update
warning: updating lock file '/workspaces/electrs/flake.lock':
• Updated input 'crane':
    'github:ipetkov/crane/a329cd00398379c62e76fc3b8d4ec2934260d636?narHash=sha256-iZDHWTqQj6z6ccqTSEOPOxQ8KMFAemInUObN2R9vHSs%3D' (2024-03-28)
  → 'github:ipetkov/crane/442a7a6152f49b907e73206dc8e1f46a61e8e873?narHash=sha256-uXNW6bapWFfkYIkK1EagydSrFMqycOYEDSq75GmUpjk%3D' (2024-05-04)
• Updated input 'nixpkgs':
    'github:NixOS/nixpkgs/2726f127c15a4cc9810843b96cad73c7eb39e443?narHash=sha256-UKcYiHWHQynzj6CN/vTcix4yd1eCu1uFdsuarupdCQQ%3D' (2024-03-27)
  → 'github:NixOS/nixpkgs/b211b392b8486ee79df6cdfb1157ad2133427a29?narHash=sha256-CLU5Tsg24Ke4%2B7sH8azHWXKd0CFd4mhLWfhYgUiDBpQ%3D' (2024-05-07)
• Updated input 'rust-overlay':
    'github:oxalica/rust-overlay/aa858717377db2ed8ffd2d44147d907baee656e5?narHash=sha256-oD4OJ3TRmVrbAuKZWxElRCyCagNCDuhfw2exBmNOy48%3D' (2024-03-28)
  → 'github:oxalica/rust-overlay/a8bfc2569a1965c0da8711d289d973f0011b441a?narHash=sha256-oujsCgNiQnZoQntNkkNkA7BhCmUvf9FLWj%2B2oGT2Jvc%3D' (2024-05-08)
warning: Git tree '/workspaces/electrs' is dirty
$ nix run .#blockstream-electrs-liquid -- --network liquid
warning: Git tree '/workspaces/electrs' is dirty
error: builder for '/nix/store/rcd2hcxdnb6v23c7mjjm33wqjbasdrvr-electrs-deps-0.4.1.drv' failed with exit code 101;
       last 25 log lines:
       >    Compiling bitcoin_hashes v0.12.0
       >     Checking object v0.32.2
       >     Checking futures-channel v0.3.30
       >     Checking which v4.4.2
       >    Compiling bitcoind v0.34.1
       >     Checking rand v0.4.6
       >     Checking iana-time-zone v0.1.60
       >     Checking tower-service v0.3.2
       >     Checking rustc-demangle v0.1.23
       >     Checking unicode-width v0.1.11
       >    Compiling bitcoin_hashes v0.10.0
       >     Checking bitcoincore-rpc-json v0.18.0
       >     Checking option-ext v0.2.0
       >     Checking winapi v0.2.8
       >    Compiling elementsd v0.9.1
       > error[E0425]: cannot find function `download_filename` in this scope
       >   --> /nix/store/cjjfa06qzk9wahmbqjh7kn86bsmaq3dv-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/elementsd-0.9.1/build.rs:59:33
       >    |
       > 59 |         let download_filename = download_filename();
       >    |                                 ^^^^^^^^^^^^^^^^^ not found in this scope
       >
       > For more information about this error, try `rustc --explain E0425`.
       >     Checking dirs-sys v0.4.1
       > error: could not compile `elementsd` (build script) due to previous error
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/rcd2hcxdnb6v23c7mjjm33wqjbasdrvr-electrs-deps-0.4.1.drv'.
error: 1 dependencies of derivation '/nix/store/z7wa0ixizb5xq32dfs8s0fv23yds4lcl-electrs-0.4.1.drv' failed to build
```